### PR TITLE
Issue #7661: Update doc for SuperClone

### LINKED ISSUE #7661: Update doc for SuperClone
https://checkstyle.org/config_coding.html#SuperClone

Add example for each existing config to show what code is ok and what code cause violation.
Good example is https://checkstyle.org/config_naming.html#AbstractClassName
Look at referenced commits of #5832, to see what and how should be changed in code.

Please play with module/check by means of https://checkstyle.org/cmdline.html to understand how it works.

------------

PLEASE READ https://github.com/checkstyle/checkstyle/wiki/Good-practice-on-how-to-update-examples

### PR DESCRIPTION
**Pull request by wilcoln**
# Resolve #7661 
example

## Ouput of default example
$ cat config.xml
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
		<module name="SuperClone"/>
	</module>
</module>
```
$ cat Test.java 
```java
class A {

 public Object clone() { // OK
  return super.clone();
 }
}

class B {
private int b;

 public B clone() { // violation, does not call super.clone()
  B other = new B();
  other.b = -1;
  return other;
 }
}

class C {

 public C clone() {  // OK
  return (C) super.clone();
 }
}
```
$ java -jar checkstyle-8.29-all.jar -c config.xml Test.java
```
Starting audit...
[ERROR] /var/tmp/Test.java:11:11: Method 'clone' should call 'super.clone'. [SuperClone]
Audit done.
Checkstyle ends with 1 errors.
```

## Ouput of non-default example
This check doesn't provide any configurable property.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 1]
```diff
@@ -37,7 +37,20 @@
  * <pre>
  * &lt;module name=&quot;SuperClone&quot;/&gt;
  * </pre>
+ * <p>Example: </p>
+ * <pre>
+ * public class Foo {
+ *
+ * public Object clone() {return null;} // violation, does not call super.clone()
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: The xdoc has the methods indented but the javadoc does not. Please make them similar so it is also easier to read in the Javadoc. The methods should be similarly indented.

@strkkk Please keep an eye out for these type of things.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 2]
```diff
@@ -37,7 +37,20 @@
  * <pre>
  * &lt;module name=&quot;SuperClone&quot;/&gt;
  * </pre>
+ * <p>Example: </p>
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: Extra space is not needed after the `:`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java (hunk 1 above)
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: It is not valid java code to have all these clones in 1 single method. Also the examples are not very creative in that they all look very similar.
What about an example where super clone is not called in the return. Like it is assigned to a variable, with some more modifications and then returned.
`clone` can't have any parameters and can't be native.

Look at the source code of the check to get an idea of some other examples that users may not think of beyond the basics.https://github.com/checkstyle/checkstyle/blob/bd45909e6345b2d3ef5c5472d35a3ad926a31eb9/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AbstractSuperCheck.java

**Comment by wilcoln** [comment]
@rnveach, is this better?

**Comment by wilcoln** [comment]
@rnveach, a review please.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 3]
```diff
@@ -37,7 +37,48 @@
...
+ * <pre>
+ * class Foo {
+ *  protected int f;
+ *
+ *  public Foo() {};
+ *  public Foo(int f) {
+ *     this.f = f;
+ *  }
+ *  // violation, does not call super.clone()
+ *  public Object clone() {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: Why can't comment go on end of line? Violation is on the method, not before or after it.

**Comment by wilcoln** [comment]
@rnveach, sorry I did not get it earlier. I fixed it properly now.

**Comment by romani** [comment]
Please update text samples in PR

**Comment by wilcoln** [comment]
> Please update text samples in PR

@romani, do you mean the `Test.java` file in the PR comment? Maybe the `Output` in the PR comment? If so both of them are up to date.
Or may be you're talking about something else?

**Comment by wilcoln** [comment]
> Please update text samples in PR

Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 4]
```diff
@@ -37,7 +37,45 @@
...
+ *  }
+ *  public Object clone() { // violation, does not call super.clone()
+ *    Foo other = new Foo();
+ *    other.f = this.f;
+ *    return other;
+ *  }
+ * }
+ *
+ * class FooA extends Foo {
+ *  private static int countA = 0;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: Please remove extra fields , Check do not care about them.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 5]
```diff
@@ -37,7 +37,45 @@
...
+ *     this.f = f;
+ *  }
+ *  public Object clone() { // violation, does not call super.clone()
+ *    Foo other = new Foo();
+ *    other.f = this.f;
+ *    return other;
+ *  }
+ * }
+ *
+ * class FooA extends Foo {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: Please extend some class from jdk, Check do not care about any hierarchy, we should not give use false feeling that consider hierarchy. All classes have such cline method to override.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 6]
```diff
@@ -37,7 +37,45 @@
...
+ *   other.a = ++countA;
+ *   return other;
+ *  }
+ * }
+ *
+ * class FooB extends Foo {
+ *  private static int countB = 0;
+ *  private int b;
+ *
+ *  public Object clone() { // violation, does not call super.clone()
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: Please show example of clone method that return exact class ( FooB ).
Please read https://github.com/checkstyle/checkstyle/issues/7553#issuecomment-590786496 and articles we reference.

**Comment by wilcoln** [comment]
@romani, a couple of days ago @rnveach commented 

> ... What about an example where super clone is not called in the return. Like it is assigned to a variable, with some more modifications and then returned...

This is the reason why I added additional fields. So what should I do?

**Comment by wilcoln** [comment]
@romani you asked me to remove extra fields, but a couple of days ago @rnveach commented 
> ... What about an example where super clone is not called in the return. Like it is assigned to a variable, with some more modifications and then returned ...
That's the reason why I introduced additional fields, to modify them before returning.
I'm a little confused, what should I do now please? keep one of them or remove all?

**Comment by wilcoln** [comment]
@romani, a review please.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java [hunk 7]
```diff
@@ -37,7 +37,32 @@
...
+ *  public Object clone() {
+ *   return super.clone(); // OK
+ *  }
+ * }
+ *
+ * class B {
+ * private int b;
  *
+ *  public B clone() {
+ *   B other = new B(); // violation, does not call super.clone()
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SuperCloneCheck.java]: Violation comment is misplaced, move it to previy line. Same for other comments.

**Comment by wilcoln** [comment]
@romani , a review plase.

**Comment by wilcoln** [comment]
@rnveach, a review please.

**Comment by wilcoln** [comment]
@romani, @rnveach, a review please.

**Comment by rnveach** [comment]
@romani
> @rnveach, please explain why you need extra fields.

Please see my request at https://github.com/checkstyle/checkstyle/pull/7718#discussion_r386565561 .

> Also the examples are not very creative in that they all look very similar.
What about an example where super clone is not called in the return. Like it is assigned to a variable, with some more modifications and then returned.

I never specifically mentioned fields. In the original review, all the examples were `return super.clone()` over and over. I wanted more originality and creativity from each example. It is not the common cases that make users think about what a new check does, it is the weird and exotic ones that show the limits of what it can and can't do.

**Comment by wilcoln** [comment]
@rnveach, message received.

**Pull request merged**